models/gif_psc_exp_multisynapse.cpp

```diff
@@ -0,0 +1,492 @@
+  assert( from < to );
+
+  for ( long_t lag = from; lag < to; ++lag )
+  {
+
+    // exponential decaying stc and sfa elements
+    S_.stc_ = 0.0;
+    for ( size_t i = 0; i < S_.stc_elems_.size(); i++ )
+    {
+      S_.stc_ += S_.stc_elems_[ i ];
+      S_.stc_elems_[ i ] = V_.P_stc_[ i ] * S_.stc_elems_[ i ];
+    }
+
+    S_.sfa_ = P_.V_T_star_;
+    for ( size_t i = 0; i < S_.sfa_elems_.size(); i++ )
+    {
+      S_.sfa_ += S_.sfa_elems_[ i ];
+      S_.sfa_elems_[ i ] = V_.P_sfa_[ i ] * S_.sfa_elems_[ i ];
+    }
+
+    double_t sum_syn_pot = 0.0;
+    for ( size_t i = 0; i < P_.num_of_receptors_; i++ )
+    {
+      // computing effect of synaptic currents on membrane potential
+      sum_syn_pot += V_.P21_syn_[ i ] * S_.i_syn_[ i ];
+      // exponential decaying PSCs
+      S_.i_syn_[ i ] = V_.P11_syn_[ i ] * S_.i_syn_[ i ];
+      S_.i_syn_[ i ] += B_.spikes_[ i ].get_value( lag ); // collecting spikes
+    }
+
+    if ( S_.r_ref_ == 0 ) // neuron not refractory, so evolve V
+    {
+      // effect of synaptic currents (sum_syn_pot) is added here
+      S_.y3_ = V_.P30_ * ( S_.y0_ + P_.I_e_ - S_.stc_ ) + V_.P33_ * S_.y3_
+        + V_.P31_ * P_.E_L_ + sum_syn_pot;
+
+      const double_t lambda =
+        P_.lambda_0_ * std::exp( ( S_.y3_ - S_.sfa_ ) / P_.Delta_V_ );
+
+      if ( lambda > 0.0 )
+      {
+        // Draw random number and compare to prob to have a spike
+        // hazard function is computed by 1 - exp(- lambda * dt)
+        if ( V_.rng_->drand()
+          < -numerics::expm1( -lambda * Time::get_resolution().get_ms() ) )
+        {
+
+          for ( size_t i = 0; i < S_.stc_elems_.size(); i++ )
+          {
+            S_.stc_elems_[ i ] += P_.q_stc_[ i ];
+          }
+
+          for ( size_t i = 0; i < S_.sfa_elems_.size(); i++ )
+          {
+            S_.sfa_elems_[ i ] += P_.q_sfa_[ i ];
+          }
+
+          S_.r_ref_ = V_.RefractoryCounts_;
+
+          // And send the spike event
+          set_spiketime( Time::step( origin.get_steps() + lag + 1 ) );
+          SpikeEvent se;
+          kernel().event_delivery_manager.send( *this, se, lag );
+        }
+      }
+    }
+    else
+    {
+      --S_.r_ref_;          // neuron is absolute refractory
+      S_.y3_ = P_.V_reset_; // reset the membrane potential
+    }
+
+    // Set new input current
+    S_.y0_ = B_.currents_.get_value( lag );
+
+    // Voltage logging
+    B_.logger_.record_data( origin.get_steps() + lag );
+  }
+}
+
+port
+gif_psc_exp_multisynapse::handles_test_event( SpikeEvent&, rport receptor_type )
+{
+  if ( receptor_type <= 0
+    || receptor_type > static_cast< port >( P_.num_of_receptors_ ) )
+    throw IncompatibleReceptorType( receptor_type, get_name(), "SpikeEvent" );
+
+  P_.has_connections_ = true;
+  return receptor_type;
+}
+
+void
+gif_psc_exp_multisynapse::handle( SpikeEvent& e )
+{
+  assert( e.get_delay() > 0 );
+
+  for ( size_t i = 0; i < P_.num_of_receptors_; ++i )
```

Comment:
I realized this code is inherited from `iaf_psc_alpha_multisynapse`, but that code is from before we instituted systematic code review. The loop here runs through all receptors and picks out exactly one in the least efficient way possible. All we need is
```C++
B_.spikes_[ e.get_rport() - 1 ].add_value( 
  e.get_rel_delivery_steps( kernel().simulation_manager.get_slice_origin() ),
  e.get_weight() * e.get_multiplicity() );
```